File: hestia/discovery.go
package hestia

import "context"

// Discovery service discovery interface
type Discovery interface {
	// GetServices returns a list of instances
	// After we obtain the service instance, we can get the currently available service instance
	// from the service list according to different strategies.
	// version 版本号参数，如果传递为空，就没有版本区分
	GetServices(ctx context.Context, name string, version string) ([]*Service, error)

	// Get returns an available service instance based on the specified service selection strategy.
	// the selection strategy is RoundRobinHandler
	Get(ctx context.Context, name string, version string, strategyHandler ...StrategyHandler) (*Service, error)

	// String returns the name of the resolver.
	String() string
}

File: rs-hestia/src/discovery.rs
use std::any::Any;

use async_trait::async_trait;

use crate::Context;
use crate::error::Result;
use crate::service::{Service, StrategyHandler};

/// Service discovery interface.
#[async_trait]
pub trait Discovery: Send + Sync + Any {
    /// Returns a list of healthy service instances.
    ///
    /// `version` may be empty to ignore version filtering.
    async fn get_services(&self, ctx: &Context, name: &str, version: &str) -> Result<Vec<Service>>;

    /// Returns a single available instance using the supplied strategy.
    ///
    /// If no strategy is supplied, round-robin is used.
    async fn get(
        &self,
        ctx: &Context,
        name: &str,
        version: &str,
        strategy: Option<StrategyHandler>,
    ) -> Result<Service>;

    /// Returns the name of the discovery implementation.
    fn name(&self) -> &str;

    /// Returns the type as `Any` so that implementations can be downcast.
    ///
    /// This mirrors Go's type assertion and is used by the gRPC resolver to
    /// reuse etcd-specific watch capabilities.
    fn as_any(&self) -> &dyn Any;
}
